src/storage/concurrency_manager/handle_table.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::key_handle::{KeyHandle, KeyHandleMutexGuard, KeyHandleRef};
+
+use kvproto::kvrpcpb::LockInfo;
+use parking_lot::Mutex;
+use std::{collections::BTreeMap, ops::Bound, sync::Arc};
+use txn_types::Key;
+
+#[derive(Default)]
+pub struct HandleTable<M>(Arc<M>);
+
+impl<M: OrderedMap> Clone for HandleTable<M> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<M: OrderedMap> HandleTable<M> {
+    pub async fn lock_key(&self, key: &Key) -> KeyHandleMutexGuard<'_, M> {
+        loop {
+            if let Some(handle) = self.0.get(key) {
+                return handle.mutex_lock().await;
+            } else {
+                let handle = Arc::new(KeyHandle::new(key.clone()));
+                let handle_ref = handle.clone().get_ref(&*self.0).unwrap();
+                let guard = handle_ref.mutex_lock().await;
+                if self.0.insert_if_not_exist(key.clone(), handle) {
+                    return guard;
+                }
+            }
+            // If the program goes here, the handle is about to be removed from the map.
+            // We should retry and get or insert the new handle.
+            // Question: should we yield here?
+        }
+    }
+
+    pub fn check_key(
+        &self,
+        key: &Key,
+        check_fn: impl FnOnce(&LockInfo) -> bool,
+    ) -> Result<(), LockInfo> {
+        if let Some(lock_ref) = self.0.get(key) {
+            return lock_ref.with_lock_info(|lock_info| {
+                if let Some(lock_info) = &*lock_info {
+                    if !check_fn(lock_info) {
+                        return Err(lock_info.clone());
+                    }
+                }
+                Ok(())
+            });
+        }
+        Ok(())
+    }
+
+    pub fn check_range(
+        &self,
+        start_key: &Key,
+        end_key: &Key,
+        mut check_fn: impl FnMut(&LockInfo) -> bool,
+    ) -> Result<(), LockInfo> {
+        let blocking_lock = self.0.find_first(start_key, end_key, |lock_ref| {
+            lock_ref.with_lock_info(|lock_info| {
+                lock_info.as_ref().and_then(|lock_info| {
+                    if check_fn(lock_info) {
+                        None
+                    } else {
+                        Some(lock_info.clone())
+                    }
+                })
+            })
+        });
+        if let Some(lock) = blocking_lock {
+            Err(lock)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// A concurrent ordered map which maps encoded keys to key handle.
+pub trait OrderedMap: Default + Send + Sync + 'static {
```

Comment:
Why do we need this abstraction? Can't we just use the concrete type? I imagine that if we change it in the future to something better, we wouldn't need to choose the backend.